### PR TITLE
fix(uniqueWith): return items that are not equal to themselves

### LINKED ISSUE #999: `uniqueWith` doesn't return items that are not equal to themselves
```ts
const errors = [
  { id: 1, reason: 'No name' },
  { id: 1, reason: 'No name' },
  { reason: 'No name' },
  { reason: 'No name' }
]

R.uniqueWith(errors, (errorA, errorB) => {
  if(errorA.id === undefined || errorB.id === undefined){
    return false
  }
  return errorA.id === errorB.id
})
```

expected
```ts
[
  { id: 1, reason: 'No name' },
  { reason: 'No name' },
  { reason: 'No name' }
]
```

actual
```ts
[
  { id: 1, reason: 'No name' },
]
```

It was unexpected for me that items not equal to themselves were removed from the array

### PR DESCRIPTION
**Pull request by tbui17**
Fixes #999 

- Previous implementation checked each item for a match that existed at a different index, skipping those. It checks the item's own equivalency as part of the implementation, but this can unexpectedly remove items due to interactions with the comparator's logic as shown in the issue.
- In the new implementation, the index equality check simultaneously handles this edge case and reduces unnecessary calls of the comparator.

---

Make sure that you:

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `docs/src/content/mapping` for both Lodash and Ramda.

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions, and changes to a function's type that would impact users.
- `fix`: changes to the runtime behavior of an existing function, or refinements to it's type that shouldn't impact most users.
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

### CODE REVIEW

File: src/uniqueWith.ts [hunk 1]
```diff
@@ -57,7 +57,16 @@ export function uniqueWith(...args: ReadonlyArray<unknown>): unknown {
 
 const lazyImplementation =
   <T>(isEquals: IsEquals<T>): LazyEvaluator<T> =>
-  (value, index, data) =>
-    data.findIndex((otherValue) => isEquals(value, otherValue)) === index
+  (value, index, data) => {
+    const otherIndex = data.findIndex((otherValue) =>
+      isEquals(value, otherValue),
+    );
```
Comment (filipw01) [inline src/uniqueWith.ts:63]: I wonder if we could optimize it like this to avoid calling `isEquals` on every item against every other Item

```
const firstEqualIndex = data.findIndex((otherValue, otherIndex) => {
      return index === otherIndex || isEquals(value, otherValue)
    });
```

File: src/uniqueWith.ts [hunk 2]
```diff
@@ -57,7 +57,16 @@ export function uniqueWith(...args: ReadonlyArray<unknown>): unknown {
 
 const lazyImplementation =
   <T>(isEquals: IsEquals<T>): LazyEvaluator<T> =>
-  (value, index, data) =>
-    data.findIndex((otherValue) => isEquals(value, otherValue)) === index
+  (value, index, data) => {
+    const otherIndex = data.findIndex((otherValue) =>
```
Comment (filipw01) [inline src/uniqueWith.ts:61]: What do you think about `firstEqualIndex` for the name of this variable instead?

**Review by filipw01 — CHANGES_REQUESTED** [review]

File: src/uniqueWith.ts (hunk 2 above)
Comment (tbui17) [reply src/uniqueWith.ts:61]: It's a lot better! The name reflects the role of the variable now. I was only mirroring the variable names before.

File: src/uniqueWith.ts (hunk 1 above)
Comment (tbui17) [reply src/uniqueWith.ts]: It's a good idea. I usually leave optimizations for a separate PR but this can definitely avoid some potentially expensive equality checks and it's a simple change.

**Review by filipw01 — APPROVED** [review]

**Review by cjquines — APPROVED** [review]
lgtm, thanks so much!
